File: backend/app/models/seating.py
from datetime import datetime, UTC
from typing import Any
from sqlalchemy import Integer, String, ForeignKey, JSON, DateTime, Boolean
from sqlalchemy.orm import Mapped, mapped_column, relationship
from app.database import Base


class Seating(Base):
    __tablename__ = "seatings"

    id: Mapped[int] = mapped_column(primary_key=True)
    class_id: Mapped[int] = mapped_column(ForeignKey("classes.id"), unique=True)
    rows: Mapped[int] = mapped_column(Integer, default=6)
    cols: Mapped[int] = mapped_column(Integer, default=8)
    seats: Mapped[list[Any]] = mapped_column(JSON, default=list)  # 2D array of student IDs
    updated_at: Mapped[datetime] = mapped_column(
        default=lambda: datetime.now(UTC),
        onupdate=lambda: datetime.now(UTC)
    )

    class_: Mapped["Class"] = relationship("Class", back_populates="seating")


class SeatingLayout(Base):
    """Named seating arrangement snapshots — one class can have many."""
    __tablename__ = "seating_layouts"

    id: Mapped[int] = mapped_column(primary_key=True)
    class_id: Mapped[int] = mapped_column(ForeignKey("classes.id"))
    name: Mapped[str] = mapped_column(String(120))
    rows: Mapped[int] = mapped_column(Integer, default=6)
    cols: Mapped[int] = mapped_column(Integer, default=8)
    seats: Mapped[list[Any]] = mapped_column(JSON, default=list)
    is_active: Mapped[bool] = mapped_column(Boolean, default=False)
    created_at: Mapped[datetime] = mapped_column(
        DateTime, default=lambda: datetime.now(UTC)
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime,
        default=lambda: datetime.now(UTC),
        onupdate=lambda: datetime.now(UTC),
    )

    class_: Mapped["Class"] = relationship("Class", back_populates="seating_layouts")

File: flutter_app/lib/data/models/seating.dart
class SeatingModel {
  final int id;
  final int classId;
  final int rows;
  final int cols;
  final List<List<int?>> seats; // 2D array of student IDs (null = empty)
  final DateTime updatedAt;

  SeatingModel({
    required this.id,
    required this.classId,
    required this.rows,
    required this.cols,
    required this.seats,
    required this.updatedAt,
  });

  factory SeatingModel.fromJson(Map<String, dynamic> json) {
    // Parse seats from JSON - could be List<dynamic> of lists
    final seatsData = json['seats'] as List<dynamic>;
    final seats = seatsData.map((row) {
      return (row as List<dynamic>).map((id) => id as int?).toList();
    }).toList();

    return SeatingModel(
      id: json['id'],
      classId: json['class_id'],
      rows: json['rows'],
      cols: json['cols'],
      seats: seats,
      updatedAt: DateTime.parse(json['updated_at']),
    );
  }

  SeatingModel copyWith({
    int? rows,
    int? cols,
    List<List<int?>>? seats,
  }) {
    return SeatingModel(
      id: id,
      classId: classId,
      rows: rows ?? this.rows,
      cols: cols ?? this.cols,
      seats: seats ?? this.seats,
      updatedAt: updatedAt,
    );
  }
}
